### PR TITLE
Lets cats attack mobs in the rat faction

### PR DESCRIPTION
**Pull request by pizie11**
## About The Pull Request

This PR will make cats target, move towards, and attack living mobs in the rat faction. Right now this includes Rats, Royal Rats, and anyone that overdosed on rat spit (>20u).

A cat's scratch will deal 4-6 damage, which will outpace a single rat, but stands no chance against a royal rat and its army.

Mouse splatting and the Tom & Jerry chase should still happen as normal, this only affects everything else in the rat faction

single cat VS. single rat

https://user-images.githubusercontent.com/6319502/191415420-a9692a90-4fe9-422a-9732-067bca9b4108.mp4

single cat VS. a royal rat

https://user-images.githubusercontent.com/6319502/191416443-ddd7b526-a300-43b3-b896-a52c37eb5943.mp4

## Why It's Good For The Game

It doesn't make any sense for cats to not at least fight back when a rat is bashing them. Rats are described as mutated mice but they are similar enough to where a cat wouldn't care about the difference and claw at it anyway. This will give them a fighting chance while also still keeping them as not a threat to royal rats.

## Changelog

:cl: Pizzie11
add: Cats will now attack beings that have kinship with rats, this includes rats, royal rats, and anyone who overdosed on rat spit.
/:cl:

### CODE REVIEW

**Comment by JohnFulpWillard** [comment]
This is a feature as regal rats aren't regular rats (they're also excluded from mice traps) and because this allows them to additionally attack humans the regal rat has converted using their rat spit.

**Comment by Valorium-RB** [comment]
Now we need royal cats

**Comment by Shadowflame909** [comment]
I guess Tom the Mouse and Jerry the cat will no longer be friends after this change

**Comment by pizie11** [comment]
> I guess Tom the Mouse and Jerry the cat will no longer be friends after this change

The Tom and Jerry interaction should still happen, I should refactor this a tad to make that more clear + fix some redundancy I just noticed

**Comment by VladinXXV** [comment]
race war starts now

File: code/modules/mob/living/simple_animal/friendly/cat.dm [hunk 1]
```diff
@@ -221,22 +223,35 @@
...
-					break
+			//Targeting anything in the rat faction nearby
+			for(var/mob/living/M in view(1,src))
 				if(!M.stat && Adjacent(M))
-					manual_emote("splats \the [M]!")
-					M.splat()
-					movement_target = null
-					stop_automated_movement = 0
-					break
+					if (faction_check(M.faction.Copy(), list("rat")))
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/friendly/cat.dm:232]: Why Copy()?

Also "rat" should be a define here and everywhere its used for typo protection

File: code/modules/mob/living/simple_animal/friendly/cat.dm [hunk 2]
```diff
@@ -221,22 +223,35 @@
...
+							snack.splat()
+							movement_target = null
+							stop_automated_movement = 0
+							break
+						//Rat scratching, or anything else that could be in the rat faction
+						//Copied from hostile simple animal code
+						if(SEND_SIGNAL(src, COMSIG_HOSTILE_PRE_ATTACKINGTARGET, M) & COMPONENT_HOSTILE_NO_ATTACK)
+							return FALSE //but more importantly return before attack_animal called
+						var/result = M.attack_animal(src)
+						SEND_SIGNAL(src, COMSIG_HOSTILE_POST_ATTACKINGTARGET, M, result)
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/friendly/cat.dm:253]: Signals should only be fired from one place.

This should instead be a shared proc that hostile code uses.

File: code/modules/mob/living/simple_animal/friendly/cat.dm [hunk 3]
```diff
@@ -257,10 +272,12 @@
...
 				movement_target = null
 				stop_automated_movement = 0
-				for(var/mob/living/simple_animal/mouse/snack in oview(src,3))
-					if(isturf(snack.loc) && !snack.stat)
-						movement_target = snack
-						break
+				//Targeting mice and mobs in the rat faction
+				for(var/mob/living/target in oview(src,3))
+					if(isturf(target.loc) && !target.stat)
+						if(faction_check(target.faction.Copy(), list("rat")))
```
Comment (Mothblocks) [inline code/modules/mob/living/simple_animal/friendly/cat.dm:278]: Why copy()?

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/friendly/cat.dm (hunk 2 above)
Comment (pizie11) [reply code/modules/mob/living/simple_animal/friendly/cat.dm:253]: Looking at this again, I didn't need to have the signals there (and copy that code in the first place) as they're mainly for hostile animal AI behavior, getting rid of everything but attack_animal still works and doesn't appear to break anything when testing.

I was also debating refactoring cats to hostile/retaliate with this PR, but I don't think that's necessary right?

File: code/modules/mob/living/simple_animal/friendly/cat.dm (hunk 1 above)
Comment (pizie11) [reply code/modules/mob/living/simple_animal/friendly/cat.dm]: Defines done and reworked the faction checking

**Pull request merged**